### ISSUE TITLE
June 19

### ISSUE THREAD

**Issue by hillarysanders** [hillarysanders/notes#45]

**Comment by hillarysanders** [comment]
sciency research
vs PE model research
--> Hillary and Maddy are going to maintain PE model, Hillary is project lead. Junior engineer --> he could be a helper there. Grouplet. --> rest of the group focuses on the research projects. 

- [ ] read more about python GIL stuff (what Josh said)

kont, rich, hillary and maddy under josh. Ethan under konst, somebody else under Rich. Lady from UMD. 
- 2 open positions - > 50% change of alex joining. Konst and Hillary liked him.  --> eng-oriented people. Filipe, alex join - then one left. William from within Sophos. Or bringing on a software engineer from another team. Increase our headcount, don't need to pay for their salaries?


- dealt with a lot of memory leaks that caused things to crash after a while - had to neuter the tensorboard results. Currently training a larger sample from dynamo table - but the issue is that a ton of the files DON'T exist in s3, which is really annoying.
- now onto validation part of the pipeline.
- string features that they were extracting - non-printable characters are split really badly.
--> now splitting on XML stuff.

The paper:
- sounds like batch norm is still necessary for ELU, but not for SELU sometimes?
low FPR ≈ signatures...
I get the vanishing gradient problem
I get that if you have an average, or all positive activation function values, then your d-error/d-weight-i will always be all negative or all positive across batches... But I'm not entirely sure how that helps. Stops zig-zaggin?

Stochastic gradient descent. Picks steepest direction.
1st order method. --> Instead, you do quadratic approximation. Solve for minimum point in the parabola, and that's the step you take. Newton step - second derivative, called a hashen.

Hashen - fisher information matrix  is closely related to the Hashen. matrix of second derivatives. Natural gradient = the solution to the quadratic problem. --> the issue in optimization - computing the Hashen is complicated and expensive, so instead we use stochastic gradient descent to approximate it. 

--> if you can make the gradient be equal to the natural gradient then you can just step in that direction. --> you want the inverse of FI to be diagonal and close to 1??

--> what follows, is that
Theorom 1: shows that FI matrix changes the gradient from original gradient to natural gradient. If this *b* is 0 then the natural gradient and gradient are equal. --> You want b to be 0. That's 0 IF the outputs of the last layer are averaged at 0.
Basically - you want the output of the activation functions to be centered around 0.

--> Figure 2: ELU average is closer to 0. In 4.1, why is leaky above RELU?
ELU allows you to move the mean towards zero... 
--> do these comparisons use batch norm? Is the only benefit that you don't need batch norm? e.g 
--> didn't get why you want mean 0. Kind of just taking it on K's word that b close to 0 gets the gradient closer to the natural gradient for some reason...
- I liked it. I was confused as to WHY you want b close to 0, still I am a little. I'd like to see comparisons between ELU, SELU and RELU, with batch norm.
- how delicate is SELU - e.g. how often will it NOT converge to 1,0?
Rich's quote:
"""
mean and variance calculated per batch than a per-observation trend towards mean-0.
"""
Scale by lambda, alpha this property holds: the self-normalizing. Symbolic solver program did it.
slight variation of dropout - you scale the weight for random samples instead of zeroing it out completely.
also need to switch activation initilization from glorot to lacoon.
- take out batch norm, replace RELU with SELU, replace dropout with new dropout, glorot to lacoon.
Tried dropping in ELU units, and they haven't improved anything for Josh, have worked for Konstantin. Really helped with training data. 
Works really well n MLP.

- what other aspects of network topologies are affected. 
- mitigates vanishing / exploding gradient problems. RELU already fixes it. 
- why an arbitrary dataset leads to these activation statistics. 
- ELU makes sense, but why does SELU not just get closer to 0, but moves completely to 0.  I guess because the gradients are like, self-normalizing because of the slopes? Why wouldn't you just do e.g. RELU with a*x, b*x, x>0

TODO 
- [ ] GET moooar html data and record how much is thrown out - also differentiate between None and parsing error.
- [ ] do validation on some different HTML datasets (e.g. time split example)
-reason why we use config and not a function is for handy defaults.
